File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
# 프로그래머스 42583 다리를 지나는 트럭 LV.2
def solution(bridge_length, weight, truck_weights):
    from collections import deque
    answer = 0
    bridge = deque([0] * bridge_length)
    trucks = deque(truck_weights)

    current_truck = 0
    while len(trucks) != 0:
        current_truck -= bridge.popleft()

        if current_truck + trucks[0] <= weight:
            current_truck += trucks[0]
            bridge.append(trucks.popleft())
        else:
            bridge.append(0)
        answer += 1

    return answer + bridge_length

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    Queue<Integer> bridge = new LinkedList<>();
    
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        makeBridge(bridge_length);
        return move(weight, truck_weights);
    }
    
    private int move(int weight, int[] trucks) {
        int limit = 0, time = 0;
        for (int truck : trucks) {
            while (true) {
                limit -= bridge.poll();
            
                if (limit + truck <= weight) {
                    bridge.add(truck);
                    limit += truck;
                    time++;
                    break;
                }
                
                bridge.add(0);
                time++;
            }
        }
        
        while (!bridge.isEmpty()) {
            bridge.poll();
            time++;
        }
        
        return time;
    }
    
    private void makeBridge(int len) {
        for (int i = 0; i < len; i++) bridge.add(0);
    }
}